kata-containers/kata-containers
# feat(runtime): vcpu resize capability

## Content
* This PR allows Kata Container 3.0(runtime-rs) to resize the vCPU on the fly, currently only working with dragonball.
  * Note that this has to work with `dragonball upcall patches` [[doc](https://github.com/kata-containers/kata-containers/blob/main/src/dragonball/docs/upcall.md)] and `dragonball upcall feature`[in Cargo.toml].
  * But to work with QEMU, I think we only need to support QMP cpu hotplugging in the runtime (TODO).

## How to test
* In the container spec, you add `cpu_quota` field and `cpu_period` field to `linux[resources]`. For more info, see [here](https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1/api.proto)
    * For example, a `cpu_quota` of `200000` and `cpu_period` of `100000` should result in 2 (200000/100000, rounded up) more vcpus in the VM.
* Set the `default_maxvcpu` field in the kata configuration toml file
* In your pod, create and start the container with the spec.
    * The resources will be updated on `Start`, `Create` and `Update`.
* You could check the cpu number with command `sudo crictl exec <container-id> cat /proc/cpuinfo` with `crictl` or similar cli tools.
